include/sparrow/array_data_factory.hpp

```diff
@@ -0,0 +1,359 @@
+        {
+            const size_t buffer_size = (values.size() * sizeof(T)) / sizeof(uint8_t);
+            array_data::buffer_type buffer(buffer_size);
+            auto data = buffer.data<T>();
+            for (const auto& value : values)
+            {
+                *data = value;
+                ++data;
+            }
+            return buffer;
+        };
+        return {
+            .type = data_descriptor(arrow_type_id<T>()),
+            .length = static_cast<int64_t>(values.size()),
+            .offset = offset,
+            .bitmap = bitmap,
+            .buffers = {create_buffer()},
+            .child_data = {},
+            .dictionary = nullptr
+        };
+    }
+
+    template <typename T>
+    array_data make_array_data_for_variable_size_binary_layout()
+    {
+        return {
+            .type = data_descriptor(arrow_type_id<T>()),
+            .length = 0,
+            .offset = 0,
+            .bitmap = {},
+            .buffers = {{}, array_data::buffer_type(sizeof(std::int64_t), 0)},
+            .child_data = {},
+            .dictionary = nullptr
+        };
+    }
+
+    template <mpl::constant_range ValueRange>
+        requires std::ranges::range<std::unwrap_ref_decay_t<std::ranges::range_value_t<ValueRange>>>
+    array_data make_array_data_for_variable_size_binary_layout(
+        ValueRange&& values,
+        const array_data::bitmap_type& bitmap,
+        std::int64_t offset
+    )
+    {
+        using U = std::unwrap_ref_decay_t<std::ranges::range_value_t<ValueRange>>;
+        if constexpr (std::ranges::range<U>)
+        {
+            using V = std::ranges::range_value_t<U>;
+            if constexpr (!is_arrow_base_type_extended<V>)
+            {
+                static_assert(false, "Unsupported range value type");
+            }
+        }
+        SPARROW_ASSERT_TRUE(values.size() == bitmap.size());
+        SPARROW_ASSERT_TRUE(std::cmp_greater_equal(values.size(), offset));
+
+        const auto create_buffers = [&values]()
+        {
+            using T = std::ranges::range_value_t<ValueRange>;
+            const auto& unwrap_value = [](const T& value)
+            {
+                if constexpr (mpl::is_reference_wrapper_v<T>)
+                {
+                    return value.get();
+                }
+                else
+                {
+                    return value;
+                }
+            };
+
+            std::vector<array_data::buffer_type> buffers(2);
+            buffers[0].resize(sizeof(std::int64_t) * (values.size() + 1), 0);
+            buffers[1].resize(std::accumulate(
+                values.begin(),
+                values.end(),
+                size_t(0),
+                [&unwrap_value](std::size_t res, const auto& s)
+                {
+                    return res + unwrap_value(s).size();
+                }
+            ));
+            auto iter = buffers[1].begin();
+            const auto offsets = buffers[0].data<std::int64_t>();
```

Comment:
- This should not be const since it will be assigned below
- The build should fail, but doe snot because the [implementation](https://github.com/man-group/sparrow/blob/main/include/sparrow/buffer.hpp#L569-L574) is missing a const

Reply:
Ah indeed, I mixed up things